### PR TITLE
Block Editor: Update Block Editor classNames to convention

### PR DESCRIPTION
**Pull request by aduth**
Previously: #14112

This nightmare of a pull request seeks to update class names for components moved from the `editor` package to `block-editor`, in order to continue respecting the [CSS class naming guidelines](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#naming) which prescribe that each component be prefixed by its package directory. It does so without breaking compatibility by _keeping_ the `.editor-` prefix as a compatibility mechanism.

In its current incarnation, it is as conservative as possible to keeping full compatibility. Unlike my [previous comment](https://github.com/WordPress/gutenberg/pull/14112#issuecomment-471575882) by which I'd jynxed myself, automated transforms are highly fraught with issues (dealing with `classnames`, string template interpolation, etc). Conversely, I chose not to get into specifics of what's "worth" keeping for compatibility, for the sake of keeping the effort short-lived and as non-contentious as possible. In both of these instances, I think revisiting in the future may be appropriate.

There is an argument to be made that we make a specific exception to the CSS naming guidelines for the `block-editor` module, allowing both `editor` and `block-editor` components to share the `editor-` prefix for the sake of avoiding this tedious transition. My cautions against this are that (a) exceptions are bad because they break expectations and (b) it introduces the risk of unintentional conflict when working with components across the two otherwise distinct modules.

**Implementation notes:**

The changes here were partly RegEx-assisted mass string replacements, with a supplementary (sometimes involved) manual curation.

Specific replacement patterns:

1. Replace string occurrences, exempting [icon slugs](https://github.com/WordPress/dashicons/tree/master/sources/svg):
   - Replace: `(['"])editor-(?!aligncenter|alignleft|alignright|bold|break|code|contract|customchar|expand|help|indent|insertmore|italic|justify|kitchensink|ltr|ol-rtl|ol|outdent|paragraph|paste-text|paste-word|quote|removeformatting|rtl|spellcheck|strikethrough|table|textcolor|ul|underline|unlink|video)([^ '"]+)`
   - With: `$1editor-$2 block-editor-$2`
   - **Note:** In initial passes, this missed some components where the string does not _begin_ with `editor-` ([example](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/block-editor/src/components/block-list/block.js#L434))
2. Remove `editor-` occurrence for functions expecting a singular argument
   - Replace: `(hasClass|contains)\( 'editor-.*? `
   - With: `$1( '`
3. Remove `editor-` occurrence for IDs
   - Replace: `(id|htmlFor|aria-.+?)(="|': ')editor-.*? `
   - With: `$1$2`
   - **Note:** There is no backwards-compatibility for IDs. They are largely used exclusively for the in-app label association.
4. Replace CSS selector occurrences for [deprecated components](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/editor/src/components/deprecated.js)
   - Replace: `\.editor-(autocomplete|alignment-toolbar|block-alignment-toolbar|block-controls|block-edit|block-format-controls|block-icon|block-inspector|block-list|block-mover|block-navigation-dropdown|block-selection-clearer|block-settings-menu|block-title|block-toolbar|color-palette|contrast-checker|copy-handler|default-block-appender|inserter|inner-blocks|inspector-advanced-controls|inspector-controls|panel-color-settings|plain-text|rich-text|media-placeholder|media-upload|multi-blocks-switcher|multi-select-scroll-into-view|navigable-toolbar|observe-typing|preserve-scroll-in-reorder|skip-to-selected-block|url-input|url-popover|warning|writing-flow)`
   - With: `.block-editor-$1`
   - **Note:** This does not cover all migrated components, only those which had been exported publicly. Additional searches for `.editor-` were manually performed (guided in some cases by "post" exception `\.editor-(?!post-)`).

Notable exceptions:

- We have some occurrences of [_clever_](https://softwareengineering.stackexchange.com/questions/25276/why-is-cleverness-considered-harmful-in-programming-by-some-people) concatenation ([example](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/block-editor/src/components/panel-color-settings/index.js#L94)). Template string interpolation in general proved tricky.
- Native components bypassed compatibility and upgraded the class names directly ([example](https://github.com/WordPress/gutenberg/blob/a06f9312633dabaa31f4559e63003a4ce00d0807/packages/block-editor/src/components/plain-text/index.native.js#L44))

**Testing Instructions:**

Give it a thorough shake, paying very close attention to visual regressions.

End-to-end tests should also reveal some issues, if present.

```
npm run test-e2e
```

Most useful review will be in considering what may have been overlooked or wrongly revised.

### CODE REVIEW

**Comment by aduth** [comment]
It took all of 20 minutes for this to receive its first merge conflict 😩

**Comment by aduth** [comment]
More notes:

- Previously mentioned patterns often didn't accommodate for XPath selectors ([example](https://github.com/WordPress/gutenberg/blob/cd456f6539981c2640f05d210782dba06a6913c8/packages/e2e-tests/specs/block-deletion.test.js#L25))
- In general, there's _lots_ of references to editor modules in `e2e-tests` and `e2e-test-utils`. Ultimately, I took to just reviewing every occurrence of the plaintext term `editor-` within the directories.

**Comment by youknowriad** [comment]
Looks like there's a valid e2e test failure but yeah let's not let this PR linger too much, so I'm approving.

**Review by youknowriad — APPROVED** [review]

**Comment by aduth** [comment]
The current failure is related to a few `editor-` class names applied outside the context of the editor or block editor modules:

https://github.com/WordPress/gutenberg/blob/589a634f6c5a0e0569d500fd387415f2ba6abfa2/packages/format-library/src/link/inline.js#L45

https://github.com/WordPress/gutenberg/blob/589a634f6c5a0e0569d500fd387415f2ba6abfa2/packages/format-library/src/link/inline.js#L62

https://github.com/WordPress/gutenberg/blob/589a634f6c5a0e0569d500fd387415f2ba6abfa2/packages/format-library/src/link/inline.js#L85

These shouldn't be assigned here, and could probably be abstracted to a component offered by `BlockEditor`. I won't assume to "correct" it here, for the purpose of keeping this short-lived, but I will update the class names to add the additional `block-editor-` prefixed class name. This will ensure consistency with what I assume is otherwise meant to extend the [`<FormatToolbar />` component](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/rich-text/format-toolbar/index.js).

cc @ellatrix

**Comment by aduth** [comment]
The above prompted me to do another search pass for the "editor-" plaintext term across all packages excluding `packages/editor`, `packages/edit-post` (I'd also previously done an audit yesterday for `e2e-tests` and `e2e-test-utils`).

There are several other erroneous occurrences:

https://github.com/WordPress/gutenberg/blob/776c0e161c27bed9d7f7e7087c68fab21479f79e/packages/components/src/positioned-at-selection/index.js#L60

https://github.com/WordPress/gutenberg/blob/776c0e161c27bed9d7f7e7087c68fab21479f79e/packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-group.js#L21

https://github.com/WordPress/gutenberg/blob/776c0e161c27bed9d7f7e7087c68fab21479f79e/packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-item.js#L93

https://github.com/WordPress/gutenberg/blob/776c0e161c27bed9d7f7e7087c68fab21479f79e/packages/edit-post/src/components/visual-editor/block-inspector-button.js#L39

https://github.com/WordPress/gutenberg/blob/776c0e161c27bed9d7f7e7087c68fab21479f79e/packages/format-library/src/image/index.js#L124

https://github.com/WordPress/gutenberg/blob/776c0e161c27bed9d7f7e7087c68fab21479f79e/packages/format-library/src/image/index.js#L147

As seen, they're all related to the format toolbar and block settings menu. 

Noted previously, `FormatToolbar` is a block-editor component, so I think it's reasonable to update all these class names to include the `block-editor-` prefix.

`editor-block-settings-menu`, however, as best I can tell never had anything to do with the editor module, but rather plugin integrations in `edit-post` ([source](https://github.com/WordPress/gutenberg/tree/master/packages/edit-post/src/components/block-settings-menu)). I expect these were either always wrongly-named or neglected as part of a previous effort to extract components out of `editor` to `edit-post`. I will not update them.

File: packages/block-editor/src/components/block-compare/index.js [hunk 1]
```diff
@@ -23,8 +23,8 @@ class BlockCompare extends Component {
 
 		return difference.map( ( item, pos ) => {
 			const classes = classnames( {
-				'editor-block-compare__added': item.added,
-				'editor-block-compare__removed': item.removed,
+				'editor-block-compare__added block-editor-block-compare__added': item.added,
```
Comment (aduth) [inline packages/block-editor/src/components/block-compare/index.js]: Of my own curiosity, I did confirm that `classnames` is capable of handling the assigning of two classes with respect to a single boolean property:

```
⇒ node       
> var cn = require( 'classnames' )
undefined
> cn( { 'foo bar': true } )
'foo bar'
```

**Comment by aduth** [comment]
Another useful pattern for auditing `block-editor` module using look-behinds, supported in the [superior](https://code.visualstudio.com/docs/editor/codebasics#_advanced-search-options) Python-based Sublime Text mass search:

```
(?<!block-)editor-
```

File: packages/block-library/src/nextpage/editor.scss [hunk 2]
```diff
@@ -1,4 +1,4 @@
-.editor-visual-editor__block[data-type="core/nextpage"] {
+.block-editor-block-list__block[data-type="core/nextpage"] {
```
Comment (aduth) [inline packages/block-library/src/nextpage/editor.scss]: Noting that this was previously very broken, `.editor-visual-editor__block` doesn't exist anywhere before or after this set of changes. It should be corrected here now.

**Review by aduth — COMMENTED** [review]
One thing we should consider in a subsequent pull request is performing an audit of excessive code line length, refactoring some of the very-long lines made yet-further-longer as of the changes here.

**Pull request merged**